vshard/storage/init.lua

```diff
@@ -3444,26 +3569,41 @@ local function storage_cfg_master_prepare(cfgctx)
             box.cfg({read_only = true})
             sync(cfgctx.new_cfg.sync_timeout)
         end
-    elseif not cfgctx.was_master_in_cfg and cfgctx.is_master_in_cfg then
+    elseif do_enable_via_config then
         --
         -- Promote does not require sync, because a replica can not have data,
         -- that is not on the current master - the replica is read only. But
         -- read_only can not be set to false here, because until box.cfg is
         -- called, it can not be guaranteed, that the promotion will be
         -- successful.
         --
-        log.info('Preparing to enable the master role')
+        log.info('Enabling master via config')
     end
 end
 
 local function storage_cfg_master_commit(cfgctx)
-    if cfgctx.was_master_in_cfg and not cfgctx.is_master_in_cfg then
-        master_on_disable()
-    elseif not cfgctx.was_master_in_cfg and cfgctx.is_master_in_cfg then
-        if cfgctx.new_cfg.read_only == nil then
-            box.cfg({read_only = false})
+    if not cfgctx.was_master_auto and not cfgctx.is_master_auto then
```

Comment:
I do realize that this code is ugly. However we can't apply your diff, because your `master_role_synchronize_internal()` relies on `box.info.ro`. When the master is configured manually, we have to make the instance a master even if it is read-only. This is where most of the complication is coming from. With auto-master we need to rely on `box.info.ro`. Otherwise just on the config.

I tried to write this code in a very dumb but easy way, that I simply consider each possible change one by one:
```Lua
if not cfgctx.was_master_auto and not cfgctx.is_master_auto then
    ...
elseif not cfgctx.was_master_auto and cfgctx.is_master_auto then
    ...
elseif cfgctx.was_master_auto and not cfgctx.is_master_auto then
    ...
else
    assert(cfgctx.was_master_auto and cfgctx.is_master_auto)
    ...
end
```
Perhaps the code could be shortened, but the cases would mix with each other. Would be hard to see what exactly happens when.

I do agree though, that it would be easier if the manually configured master would still look at `box.info.ro` and other non-master instances would simply not do that. But even more I didn't want to change the old behaviour here.

Reply:
I see, let's stay as is